mujoco_py/mjpid.pyx

```diff
@@ -24,77 +81,181 @@ cdef enum USER_DEFINED_ACTUATOR_PARAMS:
+    IDX_CAS_INTEGRAL_TIME_CONSTANT = 1,
+    IDX_CAS_INTEGRAL_MAX_CLAMP = 2,
+    IDX_CAS_PROPORTIONAL_GAIN_V = 3,
+    IDX_CAS_INTEGRAL_TIME_CONSTANT_V = 4,
+    IDX_CAS_INTEGRAL_MAX_CLAMP_V = 5,
+    IDX_CAS_EMA_SMOOTH_V = 6,
+    IDX_CAS_MAX_VEL = 7,
+
+cdef enum USER_DEFINED_CONTROLLER_DATA_CASCADE:
+    IDX_CAS_INTEGRAL_ERROR = 0,
+    IDX_CAS_INTEGRAL_ERROR_V = 1,
+    IDX_CAS_LAST_ERROR = 2,
+    IDX_CAS_LAST_ERROR_V = 3,
+    IDX_CAS_STORED_EMA_SMOOTH_V = 4,
+
+cdef mjtNum c_pi_cascade_bias(const mjModel*m, const mjData*d, int id):
+    """
+    A cascaded PID controller implementation that can control position
+    and velocity setpoints for a given actuator. To activate, set 
+    gainprm="Kp_pos Ti_pos max_clamp_pos Kp_vel Ti_vel max_clamp_vel ema_smooth_factor vel_limit" 
+    in a general type actuator in mujoco xml
+    """
     cdef double dt_in_sec = m.opt.timestep
-    cdef double error = d.ctrl[id] - d.actuator_length[id]
     cdef int NGAIN = int(const.NGAIN)
 
-    cdef double Kp = m.actuator_gainprm[id * NGAIN + IDX_PROPORTIONAL_GAIN]
-    cdef double error_deadband = m.actuator_gainprm[id * NGAIN + IDX_ERROR_DEADBAND]
-    cdef double integral_max_clamp = m.actuator_gainprm[id * NGAIN + IDX_INTEGRAL_MAX_CLAMP]
-    cdef double integral_time_const = m.actuator_gainprm[id * NGAIN + IDX_INTEGRAL_TIME_CONSTANT]
-    cdef double derivative_gain_smoothing = \
-        m.actuator_gainprm[id * NGAIN  + IDX_DERIVATIVE_GAIN_SMOOTHING]
-    cdef double derivate_time_const = m.actuator_gainprm[id * NGAIN + IDX_DERIVATIVE_TIME_CONSTANT]
+    cdef double Kp_cas = m.actuator_gainprm[id * NGAIN + IDX_CAS_PROPORTIONAL_GAIN]
 
-    cdef double effort_limit_low = m.actuator_forcerange[id * 2]
-    cdef double effort_limit_high = m.actuator_forcerange[id * 2 + 1]
+    if Kp_cas != 0:
+        # Run a position PID loop and use the result to set a desired velocity signal
+        pos_output = _pid(parameters=PIDParameters(
+            dt_seconds=dt_in_sec,
+            setpoint=d.ctrl[id],
+            feedback=d.actuator_length[id],
+            Kp=Kp_cas,
+            error_deadband=0.0,
+            integral_max_clamp=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_MAX_CLAMP],
+            integral_time_const=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_TIME_CONSTANT],
+            derivative_gain_smoothing=0.0,
+            derivative_time_const=0.0,
+            previous_errors=PIDErrors(
+                error=0.0,
+                derivative_error=0.0,
+                integral_error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR],
+            )))
 
-    if fabs(error) < error_deadband:
-        error = 0.0
+        # Save errors
+        d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_LAST_ERROR] = pos_output.errors.error
+        d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR] = pos_output.errors.integral_error
+        des_vel = Kp_cas * (pos_output.errors.error + pos_output.errors.integral_error)
```

Comment:
why des_vel only has PI no D terms? also I don't fully understand the physical meaning of des_vel now (it should be the force with "Newton" unit in PID), but on in the cascaed PID, it seems to mean velocity now.

Reply:
In this setup of cascaded PID, we're only using the PI components. There are effectively two PI controllers. First controller tracks position and outputs a velocity. Second controller tracks velocity and outputs a torque.

Reply:
I also had the question: theoretically, if we needed/wanted, could these be PIDs, or is there a reason that's a bad idea? (also gainprm limitation is there, so we'd move all params to actuator_user, which is a bigger refactor)

Reply:
also I think that des_vel is just pos_output.otput

Reply:
good point. I've updated this line as such